Sending DYK alerts to course instructors

Aeropio opened the pull request:
NOTE: Please review the pull request process before opening your first PR: https://github.com/WikiEducationFoundation/WikiEduDashboard/blob/master/CONTRIBUTING.md#pull-request-process

## What this PR does
Sends DYK alerts to course creators

## Screenshots
Before:
< add a screenshot of the UI before your change >

After:
< add a screenshot of the UI after your change >
![Screenshot from 2019-04-28 11-21-45](https://user-images.githubusercontent.com/9879210/56856879-dc6b8680-69a7-11e9-9a99-c7f44a641264.png)

bwreid commented on app/models/alert.rb:
```diff
@@ -118,6 +118,17 @@ def email_target_user
 
+  def email_course_instructor
+    return if emails_disabled?
+    return if course.nil?
```
What about combining these two lines?
```js
return if emails_disabled? || course.nil?
```

Aeropio replied:
sure that's a good idea

bwreid commented on spec/lib/alerts/dyk_nomination_monitor_spec.rb:
```diff
@@ -57,6 +61,13 @@ def mock_mailer
+      create(:courses_user, user_id: course_creator.id, course_id: course.id, role: 1)
+      allow_any_instance_of(AlertMailer).to receive(:alert).and_return(mock_mailer)
+      described_class.create_alerts_for_course_articles
+      expect(Alert.last.email_sent_at).not_to be_nil
```
A better check here might be to make sure that the person to whom the email was sent matches up with the `course_creator`. Right now, it looks like you're just checking to make sure an email was sent. I imagine you could either take a look at other properties on the `Alert.last` object or expect `AlertMailer.alert` method to receive the `course_creator`.

Aeropio replied:
ok i will check that

ragesoss commented on app/models/alert.rb:
```diff
@@ -118,6 +118,17 @@ def email_target_user
     update_attribute(:email_sent_at, Time.zone.now)
   end
 
+  def email_course_instructor
```
I just mentioned this in chat, but since we'll need a new view and mailer anyway, I think the best approach for this will be put the email logic into the mailer rather than adding it to the Alert base class and also specifying both email targets in the monitor file.

Take a look at FirstStudentAlertManager, FirstEnrolledStudentAlert, and FirstEnrolledStudentAlertMailer, and FirstEnrolledStudentAlert as examples.

Doing it that way will mean we can continue to test the monitor in basically the same way we do now, but add a new isolated test (and preview) for the mailer to check that it is emailing all the right people.

Aeropio replied:
sure i will have a look at those files and try completing the spec

Aeropio replied:
@ragesoss @bwreid when i am trying to manually call methods for sending emails from rails console, it is giving some error `Mailgun::ParseError Exception: 785: unexpected token at 'Mailgun Magnificent API'` , i suspect thats because mail gun api key is not configured because i don't have .env file in my project; how can i set the api key and domain in my development.rb file ???

Aeropio commented:
@ragesoss @bwreid  I have addressed all the comments , but i am not sure why the build is failing ??could you suggest any changes so that it builds successfully, and also please review the changes which I have done

bwreid commented on app/views/instructor_alert_mailer/email.html.haml:
```diff
@@ -0,0 +1,12 @@
+		We recommend checking the submission and encouraging your student to monitor the nomination and respond to feedback.
+		
+	%p.paragraph
+		Thank you and please let us know if you need any support.
```
Just a couple things to add to this. Please include the link to the Did You Know sheet that was in the issue.
https://github.com/WikiEducationFoundation/WikiEduDashboard/issues/2635#issuecomment-477388352

Additionally, we'll want to include some kind of signature. You can take a look at the other mailers to see how we do signatures. Let me know if you can't find it.

Aeropio replied:
@bwreid how do i get link for ` Did You Know `, i am not sure about it,

bwreid commented:
@Aeropio The errors you're having right this moment have to do with rubocop. If you run rubocop locally, you'll see the errors and be able to fix them. I added another comment, please take a look.

Aeropio commented:
> @Aeropio The errors you're having right this moment have to do with rubocop. If you run rubocop locally, you'll see the errors and be able to fix them. I added another comment, please take a look.

@bwreid i ran rubocop and fixed all the 3 offences, but my rspec is still failing

ragesoss commented:
Push up the rubocop fixes; it looks like the previous build would have passed if not for those. Then post details about what specs are still failing locally for you; we might be able to guess what's wrong.

Aeropio commented:
> Push up the rubocop fixes; it looks like the previous build would have passed if not for those. Then post details about what specs are still failing locally for you; we might be able to guess what's wrong.

ok i will push those changes

Aeropio commented:
@bwreid @ragesoss I pushed changes with the signature and also added link to Did You Know sheet mentioned in https://github.com/WikiEducationFoundation/WikiEduDashboard/issues/2635#issuecomment-477388352, please let me know if this is ok or any change needed?

ragesoss commented:
@Aeropio this should include a mailer preview, which is the easiest way to inspect and preview emails to verify that they are coming out as intended (and will let you add a screenshot to the PR). See `spec/mailers/previews` for examples of this. You can access the mailer previews from http://localhost:3000/rails/mailers

Aeropio commented:
> @Aeropio this should include a mailer preview, which is the easiest way to inspect and preview emails to verify that they are coming out as intended (and will let you add a screenshot to the PR). See `spec/mailers/previews` for examples of this. You can access the mailer previews from http://localhost:3000/rails/mailers

@ragesoss when i run `gulp build` i m getting this error
 Unable to resolve path to module 'tinymce/themes/silver/theme'  import/no-unresolved

:heavy_multiplication_x: 1 problem (1 error, 0 warnings)

[11:11:13] 'lintjs' errored after 43 s
[11:11:13] ESLintError in plugin "gulp-eslint"
Message:
   Failed with 1 error
[11:11:13] 'build' errored after 44 s
[11:11:13] ESLintError in plugin "run-sequence(lintjs)"

ragesoss commented:
@Aeropio try running `yarn --ignore-engines`. That should update the node module requirements and pull in the updated version of tinymce.

Aeropio commented:
@ragesossI am calling preview for (message to instructors which i have written ) in my local but i get this error ```
NoMethodError (undefined method `html_part' for nil:NilClass
Did you mean?  html_safe?):

``` I am not sure how to proceed, please help

bwreid commented:
@Aeropio Are you using the `.html_part` method somewhere? Can you share how it's being used?

Aeropio commented:
> @Aeropio Are you using the `.html_part` method somewhere? Can you share how it's being used?

@bwreid  I am not using `.html_part` anywhere

bwreid commented:
@Aeropio This is a bit difficult to reproduce at the moment. If you can share a bit more about how you're getting this error I can try and help.

Aeropio commented:
> @Aeropio This is a bit difficult to reproduce at the moment. If you can share a bit more about how you're getting this error I can try and help.

@bwreid When i run guard(or rails s), server is starting but if go to preview any mailer ex: http://localhost:3000/rails/mailers/active_course_mailer/submission then I am getting `FATAL] NoMethodError (undefined method `html_safe' for #<ActionMailer::MessageDelivery:0x00007fcf2167ad98>
Did you mean?  html_safe?
               html_part):` in the console and simultaneously for the gulp i am getting `WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
Entrypoints:
  main (1.55 MiB)
      main.d6a7036793845521927d.js
  styleguide (595 KiB)
      styleguide.d6a7036793845521927d.js
  charts (415 KiB)
      charts.d6a7036793845521927d.js
  tinymce (798 KiB)
      tinymce.d6a7036793845521927d.js
WARNING in webpack performance recommendations: 
You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
For more info visit https://webpack.js.org/guides/code-splitting/`.

ragesoss commented:
@Aeropio that's a new error to me. Is there a stack trace that goes with it? Remind me what operating system you're using?

The `gulp` warnings are expected, and aren't related to the problem.

Aeropio commented:
> @Aeropio that's a new error to me. Is there a stack trace that goes with it? Remind me what operating system you're using?
> 
> The `gulp` warnings are expected, and aren't related to the problem.

@ragesoss  I am using ubuntu (18.04.1 LTS) operating system.The stack trace is as follows
`Started GET "/rails/mailers" for ::1 at 2019-04-11 19:08:20 +1000
[2019-04-11 19:08:20.578 DEBUG]  (38.7ms)  SET NAMES utf8,  @@SESSION.sql_mode = CONCAT(CONCAT(@@sql_mode, ',STRICT_ALL_TABLES'), ',NO_AUTO_VALUE_ON_ZERO'),  @@SESSION.sql_auto_is_null = 0, @@SESSION.wait_timeout = 2147483
[2019-04-11 19:08:20.578 DEBUG] ↳ /home/azrapc/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/activerecord-5.2.3/lib/active_record/log_subscriber.rb:98
[2019-04-11 19:08:20.712 DEBUG]  (1.5ms)  SELECT `schema_migrations`.`version` FROM `schema_migrations` ORDER BY `schema_migrations`.`version` ASC
[2019-04-11 19:08:20.713 DEBUG] ↳ /home/azrapc/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/activerecord-5.2.3/lib/active_record/log_subscriber.rb:98
[2019-04-11 19:08:24.319 INFO ] Processing by Rails::MailersController#index as HTML
[2019-04-11 19:08:24.497 INFO ] Rendering /home/azrapc/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/railties-5.2.3/lib/rails/templates/layouts/application.html.erb
[2019-04-11 19:08:24.500 INFO ] Rendering /home/azrapc/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/railties-5.2.3/lib/rails/templates/rails/mailers/index.html.erb within layouts/application
[2019-04-11 19:08:24.527 INFO ] Rendered /home/azrapc/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/railties-5.2.3/lib/rails/templates/rails/mailers/index.html.erb within layouts/application (27.5ms)
[2019-04-11 19:08:24.528 INFO ] Rendered /home/azrapc/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/railties-5.2.3/lib/rails/templates/layouts/application.html.erb (30.5ms)
[2019-04-11 19:08:24.529 INFO ] Completed 200 OK in 209ms (Views: 45.5ms | ActiveRecord: 0.0ms)
[2019-04-11 19:08:36.949 INFO ] Started GET "/rails/mailers/active_course_mailer/submission" for ::1 at 2019-04-11 19:08:36 +1000
[2019-04-11 19:08:36.985 INFO ] Processing by Rails::MailersController#preview as HTML
[2019-04-11 19:08:36.985 INFO ] Parameters: {"path"=>"active_course_mailer/submission"}
[2019-04-11 19:08:37.271 DEBUG] Course Load (0.8ms)  SELECT  `courses`.* FROM `courses` ORDER BY `courses`.`id` DESC LIMIT 1
[2019-04-11 19:08:37.271 DEBUG] ↳ spec/mailers/previews/active_course_mailer_preview.rb:5
[2019-04-11 19:08:37.487 INFO ] Rendering layouts/mailer.html.haml
[2019-04-11 19:08:37.489 INFO ] Rendering active_course_mailer/email.html.haml within layouts/mailer
[2019-04-11 19:08:37.514 INFO ] Rendered active_course_mailer/email.html.haml within layouts/mailer (25.5ms)
[2019-04-11 19:08:37.532 INFO ] Rendered layouts/mailer.html.haml (44.2ms)
[2019-04-11 19:08:37.755 DEBUG] ActiveCourseMailer#email: processed outbound mail in 296.5ms
[2019-04-11 19:08:37.792 INFO ] Completed 500 Internal Server Error in 807ms (ActiveRecord: 70.2ms)
[2019-04-11 19:08:38.498 FATAL] 
[2019-04-11 19:08:38.498 FATAL] NoMethodError (undefined method `html_safe' for #<ActionMailer::MessageDelivery:0x000055ab134eb268>
Did you mean?  html_safe?
               html_part):
[2019-04-11 19:08:38.498 FATAL] 
[2019-04-11 19:08:38.498 FATAL] /home/azrapc/.rbenv/versions/2.5.0/lib/ruby/2.5.0/delegate.rb:87:in `method_missing'
actionmailer (5.2.3) lib/action_mailer/inline_preview_interceptor.rb:48:in `html_part'
actionmailer (5.2.3) lib/action_mailer/inline_preview_interceptor.rb:29:in `transform!'
actionmailer (5.2.3) lib/action_mailer/inline_preview_interceptor.rb:21:in `previewing_email'
actionmailer (5.2.3) lib/action_mailer/preview.rb:121:in `block in inform_preview_interceptors'
actionmailer (5.2.3) lib/action_mailer/preview.rb:120:in `each'
actionmailer (5.2.3) lib/action_mailer/preview.rb:120:in `inform_preview_interceptors'
actionmailer (5.2.3) lib/action_mailer/preview.rb:75:in `call'
railties (5.2.3) lib/rails/mailers_controller.rb:28:in `preview'
actionpack (5.2.3) lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'
actionpack (5.2.3) lib/abstract_controller/base.rb:194:in `process_action'
actionpack (5.2.3) lib/action_controller/metal/rendering.rb:30:in `process_action'
actionpack (5.2.3) lib/abstract_controller/callbacks.rb:42:in `block in process_action'
activesupport (5.2.3) lib/active_support/callbacks.rb:109:in `block in run_callbacks'
sentry-raven (2.9.0) lib/raven/integrations/rails/controller_transaction.rb:7:in `block in included'
activesupport (5.2.3) lib/active_support/callbacks.rb:118:in `instance_exec'
activesupport (5.2.3) lib/active_support/callbacks.rb:118:in `block in run_callbacks'
activesupport (5.2.3) lib/active_support/callbacks.rb:136:in `run_callbacks'
actionpack (5.2.3) lib/abstract_controller/callbacks.rb:41:in `process_action'
actionpack (5.2.3) lib/action_controller/metal/rescue.rb:22:in `process_action'
actionpack (5.2.3) lib/action_controller/metal/instrumentation.rb:34:in `block in process_action'
activesupport (5.2.3) lib/active_support/notifications.rb:168:in `block in instrument'
activesupport (5.2.3) lib/active_support/notifications/instrumenter.rb:23:in `instrument'
activesupport (5.2.3) lib/active_support/notifications.rb:168:in `instrument'
actionpack (5.2.3) lib/action_controller/metal/instrumentation.rb:32:in `process_action'
actionpack (5.2.3) lib/action_controller/metal/params_wrapper.rb:256:in `process_action'
activerecord (5.2.3) lib/active_record/railties/controller_runtime.rb:24:in `process_action'
actionpack (5.2.3) lib/abstract_controller/base.rb:134:in `process'
actionview (5.2.3) lib/action_view/rendering.rb:32:in `process'
actionpack (5.2.3) lib/action_controller/metal.rb:191:in `dispatch'
actionpack (5.2.3) lib/action_controller/metal.rb:252:in `dispatch'
actionpack (5.2.3) lib/action_dispatch/routing/route_set.rb:52:in `dispatch'
actionpack (5.2.3) lib/action_dispatch/routing/route_set.rb:34:in `serve'
actionpack (5.2.3) lib/action_dispatch/journey/router.rb:52:in `block in serve'
actionpack (5.2.3) lib/action_dispatch/journey/router.rb:35:in `each'
actionpack (5.2.3) lib/action_dispatch/journey/router.rb:35:in `serve'
actionpack (5.2.3) lib/action_dispatch/routing/route_set.rb:840:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/action_dispatch/routing/route_set.rb:14:in `block in call'
skylight-core (3.1.5) lib/skylight/core/fanout.rb:25:in `instrument'
skylight-core (3.1.5) lib/skylight/core/probes/action_dispatch/routing/route_set.rb:13:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
omniauth (1.9.0) lib/omniauth/strategy.rb:192:in `call!'
omniauth (1.9.0) lib/omniauth/strategy.rb:169:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
omniauth (1.9.0) lib/omniauth/strategy.rb:192:in `call!'
omniauth (1.9.0) lib/omniauth/strategy.rb:169:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/rack/agent_hooks.rb:30:in `traced_call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/rack/browser_monitoring.rb:32:in `traced_call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
http_accept_language (2.1.1) lib/http_accept_language/middleware.rb:14:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
i18n-js (3.2.1) lib/i18n/js/middleware.rb:14:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
i18n-js (3.2.1) lib/i18n/js/middleware.rb:14:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
warden (1.2.8) lib/warden/manager.rb:36:in `block in call'
warden (1.2.8) lib/warden/manager.rb:34:in `catch'
warden (1.2.8) lib/warden/manager.rb:34:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/tempfile_reaper.rb:15:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/etag.rb:25:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/conditional_get.rb:25:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/head.rb:12:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/http/content_security_policy.rb:18:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/session/abstract/id.rb:232:in `context'
rack (2.0.6) lib/rack/session/abstract/id.rb:226:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/cookies.rb:670:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
activerecord (5.2.3) lib/active_record/migration.rb:559:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/callbacks.rb:28:in `block in call'
activesupport (5.2.3) lib/active_support/callbacks.rb:98:in `run_callbacks'
actionpack (5.2.3) lib/action_dispatch/middleware/callbacks.rb:26:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/executor.rb:14:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/debug_exceptions.rb:61:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
railties (5.2.3) lib/rails/rack/logger.rb:38:in `call_app'
railties (5.2.3) lib/rails/rack/logger.rb:26:in `block in call'
activesupport (5.2.3) lib/active_support/tagged_logging.rb:71:in `block in tagged'
activesupport (5.2.3) lib/active_support/tagged_logging.rb:28:in `tagged'
activesupport (5.2.3) lib/active_support/tagged_logging.rb:71:in `tagged'
railties (5.2.3) lib/rails/rack/logger.rb:26:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
request_store (1.4.1) lib/request_store/middleware.rb:19:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/request_id.rb:27:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/action_dispatch/request_id.rb:12:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/method_override.rb:22:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/runtime.rb:22:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
activesupport (5.2.3) lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/executor.rb:14:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/static.rb:127:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/sendfile.rb:111:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack-cors (1.0.3) lib/rack/cors.rb:95:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
sentry-raven (2.9.0) lib/raven/integrations/rack.rb:51:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
railties (5.2.3) lib/rails/engine.rb:524:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
puma (3.12.1) lib/puma/configuration.rb:227:in `call'
puma (3.12.1) lib/puma/server.rb:660:in `handle_request'
puma (3.12.1) lib/puma/server.rb:474:in `process_client'
puma (3.12.1) lib/puma/server.rb:334:in `block in run'
puma (3.12.1) lib/puma/thread_pool.rb:135:in `block in spawn_thread'
[2019-04-11 19:08:38.520 INFO ] Processing by ErrorsController#internal_server_error as HTML
[2019-04-11 19:08:38.520 INFO ] Parameters: {"path"=>"active_course_mailer/submission"}
[2019-04-11 19:08:38.609 INFO ] Rendering layouts/application.html.haml
[2019-04-11 19:08:38.612 INFO ] Rendering errors/internal_server_error.html.haml within layouts/application
[2019-04-11 19:08:38.617 INFO ] Rendered errors/internal_server_error.html.haml within layouts/application (4.6ms)
[2019-04-11 19:08:38.690 INFO ] Rendered layouts/application.html.haml (80.5ms)
[2019-04-11 19:08:38.744 INFO ] Rendering text template
[2019-04-11 19:08:38.744 INFO ] Rendered text template (0.1ms)
[2019-04-11 19:08:38.745 INFO ] Completed 500 Internal Server Error in 224ms (Views: 54.1ms | ActiveRecord: 0.0ms)
[2019-04-11 19:08:54.869 INFO ] Started GET "/rails/mailers/active_course_mailer/submission" for ::1 at 2019-04-11 19:08:54 +1000
[2019-04-11 19:08:54.899 INFO ] Processing by Rails::MailersController#preview as HTML
[2019-04-11 19:08:54.900 INFO ] Parameters: {"path"=>"active_course_mailer/submission"}
[2019-04-11 19:08:54.904 DEBUG] Course Load (0.5ms)  SELECT  `courses`.* FROM `courses` ORDER BY `courses`.`id` DESC LIMIT 1
[2019-04-11 19:08:54.904 DEBUG] ↳ spec/mailers/previews/active_course_mailer_preview.rb:5
[2019-04-11 19:08:54.913 INFO ] Rendering layouts/mailer.html.haml
[2019-04-11 19:08:54.914 INFO ] Rendering active_course_mailer/email.html.haml within layouts/mailer
[2019-04-11 19:08:54.941 INFO ] Rendered active_course_mailer/email.html.haml within layouts/mailer (26.6ms)
[2019-04-11 19:08:54.988 INFO ] Rendered layouts/mailer.html.haml (74.8ms)
[2019-04-11 19:08:54.990 DEBUG] ActiveCourseMailer#email: processed outbound mail in 84.0ms
[2019-04-11 19:08:55.397 INFO ] Completed 500 Internal Server Error in 102ms (ActiveRecord: 0.5ms)
[2019-04-11 19:08:55.135 FATAL] 
[2019-04-11 19:08:55.135 **FATAL] NoMethodError (undefined method `html_safe' for #<ActionMailer::MessageDelivery:0x000055ab11134680>
Did you mean?  html_safe?
               html_part):
[2019-04-11 19:08:55.136 FATAL] 
[2019-04-11 19:08:55.136 FATAL] /home/azrapc/.rbenv/versions/2.5.0/lib/ruby/2.5.0/delegate.rb:87:in `method_missing'**
actionmailer (5.2.3) lib/action_mailer/inline_preview_interceptor.rb:48:in `html_part'
actionmailer (5.2.3) lib/action_mailer/inline_preview_interceptor.rb:29:in `transform!'
actionmailer (5.2.3) lib/action_mailer/inline_preview_interceptor.rb:21:in `previewing_email'
actionmailer (5.2.3) lib/action_mailer/preview.rb:121:in `block in inform_preview_interceptors'
actionmailer (5.2.3) lib/action_mailer/preview.rb:120:in `each'
actionmailer (5.2.3) lib/action_mailer/preview.rb:120:in `inform_preview_interceptors'
actionmailer (5.2.3) lib/action_mailer/preview.rb:75:in `call'
railties (5.2.3) lib/rails/mailers_controller.rb:28:in `preview'
actionpack (5.2.3) lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'
actionpack (5.2.3) lib/abstract_controller/base.rb:194:in `process_action'
actionpack (5.2.3) lib/action_controller/metal/rendering.rb:30:in `process_action'
actionpack (5.2.3) lib/abstract_controller/callbacks.rb:42:in `block in process_action'
activesupport (5.2.3) lib/active_support/callbacks.rb:109:in `block in run_callbacks'
sentry-raven (2.9.0) lib/raven/integrations/rails/controller_transaction.rb:7:in `block in included'
activesupport (5.2.3) lib/active_support/callbacks.rb:118:in `instance_exec'
activesupport (5.2.3) lib/active_support/callbacks.rb:118:in `block in run_callbacks'
activesupport (5.2.3) lib/active_support/callbacks.rb:136:in `run_callbacks'
actionpack (5.2.3) lib/abstract_controller/callbacks.rb:41:in `process_action'
actionpack (5.2.3) lib/action_controller/metal/rescue.rb:22:in `process_action'
actionpack (5.2.3) lib/action_controller/metal/instrumentation.rb:34:in `block in process_action'
activesupport (5.2.3) lib/active_support/notifications.rb:168:in `block in instrument'
activesupport (5.2.3) lib/active_support/notifications/instrumenter.rb:23:in `instrument'
activesupport (5.2.3) lib/active_support/notifications.rb:168:in `instrument'
actionpack (5.2.3) lib/action_controller/metal/instrumentation.rb:32:in `process_action'
actionpack (5.2.3) lib/action_controller/metal/params_wrapper.rb:256:in `process_action'
activerecord (5.2.3) lib/active_record/railties/controller_runtime.rb:24:in `process_action'
actionpack (5.2.3) lib/abstract_controller/base.rb:134:in `process'
actionview (5.2.3) lib/action_view/rendering.rb:32:in `process'
actionpack (5.2.3) lib/action_controller/metal.rb:191:in `dispatch'
actionpack (5.2.3) lib/action_controller/metal.rb:252:in `dispatch'
actionpack (5.2.3) lib/action_dispatch/routing/route_set.rb:52:in `dispatch'
actionpack (5.2.3) lib/action_dispatch/routing/route_set.rb:34:in `serve'
actionpack (5.2.3) lib/action_dispatch/journey/router.rb:52:in `block in serve'
actionpack (5.2.3) lib/action_dispatch/journey/router.rb:35:in `each'
actionpack (5.2.3) lib/action_dispatch/journey/router.rb:35:in `serve'
actionpack (5.2.3) lib/action_dispatch/routing/route_set.rb:840:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/action_dispatch/routing/route_set.rb:14:in `block in call'
skylight-core (3.1.5) lib/skylight/core/fanout.rb:25:in `instrument'
skylight-core (3.1.5) lib/skylight/core/probes/action_dispatch/routing/route_set.rb:13:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
omniauth (1.9.0) lib/omniauth/strategy.rb:192:in `call!'
omniauth (1.9.0) lib/omniauth/strategy.rb:169:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
omniauth (1.9.0) lib/omniauth/strategy.rb:192:in `call!'
omniauth (1.9.0) lib/omniauth/strategy.rb:169:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/rack/agent_hooks.rb:30:in `traced_call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/rack/browser_monitoring.rb:32:in `traced_call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
http_accept_language (2.1.1) lib/http_accept_language/middleware.rb:14:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
i18n-js (3.2.1) lib/i18n/js/middleware.rb:14:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
i18n-js (3.2.1) lib/i18n/js/middleware.rb:14:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
warden (1.2.8) lib/warden/manager.rb:36:in `block in call'
warden (1.2.8) lib/warden/manager.rb:34:in `catch'
warden (1.2.8) lib/warden/manager.rb:34:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/tempfile_reaper.rb:15:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/etag.rb:25:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/conditional_get.rb:25:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/head.rb:12:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/http/content_security_policy.rb:18:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/session/abstract/id.rb:232:in `context'
rack (2.0.6) lib/rack/session/abstract/id.rb:226:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/cookies.rb:670:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
activerecord (5.2.3) lib/active_record/migration.rb:559:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/callbacks.rb:28:in `block in call'
activesupport (5.2.3) lib/active_support/callbacks.rb:98:in `run_callbacks'
actionpack (5.2.3) lib/action_dispatch/middleware/callbacks.rb:26:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/executor.rb:14:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/debug_exceptions.rb:61:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
railties (5.2.3) lib/rails/rack/logger.rb:38:in `call_app'
railties (5.2.3) lib/rails/rack/logger.rb:26:in `block in call'
activesupport (5.2.3) lib/active_support/tagged_logging.rb:71:in `block in tagged'
activesupport (5.2.3) lib/active_support/tagged_logging.rb:28:in `tagged'
activesupport (5.2.3) lib/active_support/tagged_logging.rb:71:in `tagged'
railties (5.2.3) lib/rails/rack/logger.rb:26:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
request_store (1.4.1) lib/request_store/middleware.rb:19:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/request_id.rb:27:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/action_dispatch/request_id.rb:12:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/method_override.rb:22:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/runtime.rb:22:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
activesupport (5.2.3) lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/executor.rb:14:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
actionpack (5.2.3) lib/action_dispatch/middleware/static.rb:127:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack (2.0.6) lib/rack/sendfile.rb:111:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
rack-cors (1.0.3) lib/rack/cors.rb:95:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
sentry-raven (2.9.0) lib/raven/integrations/rack.rb:51:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
skylight-core (3.1.5) lib/skylight/core/probes/middleware.rb:26:in `call'
railties (5.2.3) lib/rails/engine.rb:524:in `call'
newrelic_rpm (6.2.0.354) lib/new_relic/agent/instrumentation/middleware_tracing.rb:99:in `call'
puma (3.12.1) lib/puma/configuration.rb:227:in `call'
puma (3.12.1) lib/puma/server.rb:660:in `handle_request'
puma (3.12.1) lib/puma/server.rb:474:in `process_client'
puma (3.12.1) lib/puma/server.rb:334:in `block in run'
puma (3.12.1) lib/puma/thread_pool.rb:135:in `block in spawn_thread'
[2019-04-11 19:08:55.144 INFO ] Processing by ErrorsController#internal_server_error as HTML
[2019-04-11 19:08:55.144 INFO ] Parameters: {"path"=>"active_course_mailer/submission"}
[2019-04-11 19:08:55.152 INFO ] Rendering layouts/application.html.haml
[2019-04-11 19:08:55.156 INFO ] Rendering errors/internal_server_error.html.haml within layouts/application
[2019-04-11 19:08:55.188 INFO ] Rendered errors/internal_server_error.html.haml within layouts/application (31.3ms)
[2019-04-11 19:08:55.205 INFO ] Rendered layouts/application.html.haml (50.5ms)
[2019-04-11 19:08:55.205 INFO ] Rendering text template
[2019-04-11 19:08:55.205 INFO ] Rendered text template (0.0ms)
[2019-04-11 19:08:55.206 INFO ] Completed 500 Internal Server Error in 61ms (Views: 0.7ms | ActiveRecord: 0.0ms)`

ragesoss commented:
Hmm... from googling this, my best guess is that the preview code ends up calling `deliver_now` on a mailer instance.

If that's not it, push up your preview code and I can try it out locally.

Aeropio commented:
@ragesoss I have attached the screenshot of UI and pushed the changes to this branch, please review this

ragesoss commented on app/mailers/instructor_alert_mailer.rb:
```diff
@@ -0,0 +1,22 @@
+
+class InstructorAlertMailer < ApplicationMailer
+  def email(alert)
+    # return unless Features.email?
```
This should not be commented out.

ragesoss replied:
To get around the problem of being able to preview the mail, even with `Features.email?` off, see how some of the other mailers are structured, for example, WikiEmailMailer. The idea is to use a class method to call it from other application code, and have the `email` method only called either from the preview or from that class method. That way, we still get the guard clause based on whether email is enabled (and whatever else needs it), and we can still get a working preview.

Aeropio replied:
yes ok i will check this

ragesoss commented on app/mailers/instructor_alert_mailer.rb:
```diff
@@ -0,0 +1,22 @@
+    # return unless Features.email?
+    @alert = alert
+    set_course_and_creators
+    if @instructors.present?
```
This is better as a guard clause: `return unless @instructors.any?` or `return if @instructors.empty?`

Aeropio replied:
sure

ragesoss commented:
Sorry for the delay. I remember starting to review this before, and then I must have gotten distracted before I left comments.

I think the email looks solid. A couple changes I can suggest, in addition to the inline comments:

* for the link to the PDF handout, put do it like "Did You Know nomination ([PDF handout](https://upload.wikimedia.org/wikipedia/commons/e/ee/Classroom_handout_-_Submitting_an_article_to_the_Did_You_Know_process.pdf))"
* for the preview, don't create a new record every time, if there's a relevant record already in the database.
* the mailer class needs a better names, such `DidYouKnowAlertMailer`, so that we can easily remember what that class if for.

Aeropio commented:
> Sorry for the delay. I remember starting to review this before, and then I must have gotten distracted before I left comments.
> 
> I think the email looks solid. A couple changes I can suggest, in addition to the inline comments:
> 
> * for the link to the PDF handout, put do it like "Did You Know nomination ([PDF handout](https://upload.wikimedia.org/wikipedia/commons/e/ee/Classroom_handout_-_Submitting_an_article_to_the_Did_You_Know_process.pdf))"
> * for the preview, don't create a new record every time, if there's a relevant record already in the database.
> * the mailer class needs a better names, such `DidYouKnowAlertMailer`, so that we can easily remember what that class if for.

I have changes the name of the file and made other inline changes, @ragesoss please have a look

ragesoss commented on app/views/did_you_know_alert_mailer/email.html.haml:
```diff
@@ -0,0 +1,19 @@
+  Hello,
+  %p.paragraph
+    The following article being worked on as a part of your course has received a Did You Know nomination
+    %a.link{href: "https://upload.wikimedia.org/wikipedia/commons/e/ee/Classroom_handout_-_Submitting_an_article_to_the_Did_You_Know_process.pdf"}= "#{"(PDF handout)"}: "
```
Just "PDF handout" should be a link, not the parens or colon.

Looking at the new screenshot, I think it will be better to reorganize the text a little so that it's more clear that the article is the important part, and the PDF is for context. I suggest:

> An article that is being worked on as part of your course, `<article link>`, has received a Did You Know nomination. See the [PDF handout about the Did You Know process](https://upload.wikimedia.org/wikipedia/commons/e/ee/Classroom_handout_-_Submitting_an_article_to_the_Did_You_Know_process.pdf) for more details.

Aeropio replied:
ok

ragesoss commented on spec/mailers/previews/did_you_know_alert_mailer_preview.rb:
```diff
@@ -0,0 +1,15 @@
+
+class DidYouKnowAlertMailerPreview < ActionMailer::Preview
+  def message_to_instructors
+    dk_nomination_alert = Alert.where(type: 'DYKNominationAlert')
```
You can do `Alert.find_by` here instead, to fetch the first record that matches.

Aeropio commented:
@ragesoss updated the UI and resolved all comments, please take a look

ragesoss commented on app/views/did_you_know_alert_mailer/email.html.haml:
```diff
@@ -0,0 +1,20 @@
+  Hello,
+  %p.paragraph
+    An article that is being worked on as part of your course,
+    %a.link{href: @alert.article.url}= @alert.article.title
```
This needs a trailing comma.

ragesoss commented on app/views/did_you_know_alert_mailer/email.html.haml:
```diff
@@ -0,0 +1,20 @@
+    An article that is being worked on as part of your course,
+    %a.link{href: @alert.article.url}= @alert.article.title
+    has received a Did You Know nomination.
+    %a.link{href: "https://upload.wikimedia.org/wikipedia/commons/e/ee/Classroom_handout_-_Submitting_an_article_to_the_Did_You_Know_process.pdf"}= "#{"See the PDF handout about the Did You Know process for more details."} "
```
Too much of the text is part of the link here. I think just linking the "PDF handout about the Did You Know process" part works better.

It's also not necessary to use string interpolation here. If there are no variables involved, you can do
```haml
%a.link{href: foo}
  Link text
```

Aeropio replied:
yes ok

Aeropio commented:
@rageross updated the UI and addressed the comments
